### DISCUSSION TITLE
Revalidate array of paths – On demand revalidate

### DISCUSSION THREAD

**Discussion by heggemsnes** [vercel/next.js#34567] [Ideas]
### Describe the feature you'd like to request

It would be great if unstable_revalidate could accept an array of paths. 

A lot of time content is referenced on other pages. By only updating the main page via i.e. a webhook, stale content can be shown on other pages. 

I.e. updating the phone number of a contact person. We both want to update the /person/<name> but we also might want to update pages where the person is referenced such as /solutions/finance.

### Describe the solution you'd like

Allow array of paths in unstable_revalidate.

``await res.unstable_revalidate(['/person/john/', '/news/article-written-by-tom'])``

### Describe alternatives you've considered

````
// /api/revalidate.ts

 Promise.all(
    paths.map(async (path) => {
      await res.unstable_revalidate(path)
    })
  )
````

This should work, but is probably not very scaleable.

**Comment by musme** [comment]
I also think this is a very necessary feature!🙏

**Comment by IamManchanda** [comment]
Came here from https://github.com/vercel/next.js/issues/34581 ... I Upvote this feature ... It is a MUST

Currently

```ts
unstable_revalidate: (urlPath: string) => Promise<void>;
revalidate: (urlPath: string) => Promise<void>; // stable version
```

What is Needed...

```ts
unstable_revalidate: (urlPaths: string | string[]) => Promise<void>;
revalidate: (urlPaths: string | string[]) => Promise<void>; // stable version
```

or maybe a `revalidateAll` or `revalidateMany` function  ... whatever seems better to the authors ... thanks 😃

**Comment by martinnabhan** [comment]
Great suggestion!
I believe not only do arrays have to be supported, but there should be a mechanism for revalidating dynamic pages without exactly specifying every path, as I've outlined here: https://github.com/vercel/next.js/discussions/34585.

**Comment by utkarshk384** [comment]
Does it currently support dynamic page revalidation? Since I tried doing it on a dynamic page and got slammed with this XML error

```
XML Parsing Error: syntax error
Location: http://localhost:4000/api/revalidate
Line Number 1, Column 1:
```

Has anyone tried it with dynamic routing?

Edit: As a temporary solution I think you can do the following:

```typescript
//Component.ts

const pages = ['/person/john/', '/news/article-written-by-tom']

pages?.forEach(async (item) => {
      const config: AxiosRequestConfig = {
        method: "post",
        url: "/api/revalidate",
        headers: {
          "Content-Type": "application/json",
          "token": process.env.TOKEN,
        },
        data: { revalidationURL: item },
      };

      const response = await axios.request(config);

      if (response.status !== 200)
        console.warn({
          message: `Error in revalidating the page ${item}`,
          status: response.status,
        });
    });

```

```typescript
// revalidate.ts

import { NextApiHandler } from "next/types";

const handler: NextApiHandler = async (req, res) => {
  // Check for secret to confirm this is a valid request
  const token = req.headers["token"];
  if (token !== process.env.TOKEN) {
    return res.status(401).json({ message: "Invalid token" });
  }

  try {

    await res.unstable_revalidate(`/testimonials/${req.body.revalidationURL}`);
    return res.status(200).json({ revalidated: true });
  } catch (err) {

    console.error(`Error revalidating page ${req.body.revalidationURL}: ${err}`);
    return res.status(500).send("Error revalidating");
  }
};

export default handler;
```

In order to avoid exposing the revalidation token to the front-end(`NEXT_PUBLIC_TOKEN`) you can send the `TOKEN` along with your `getStaticProps` at build time.

Note: This method does have some consequences. 
1. Since this method is calling the API endpoint again and again; there might be a chance that it might cost you a lot more when you're using it in a serverless environment(not tested, just my guess). 
2. It also might be slow if you are using `HTTP/1.1` and have a lot of pages to revalidate since HTTP/1.1 only allows `6` concurrent requests at a time.

**Reply by kag359six** [reply]
Not 100% sure I followed what you're asking, but since dynamic pages export a `getStaticPaths` function, you can reuse that same function since it's what Next.js would call to generate your pages anyways. Here's an example:

```
// in your api/revalidate.js file

import { getStaticPaths } from 'pages/articles/[slug]'

export default async function revalidate(req, res) {
  try {
    // grab static paths using the same method Next.js would use
    const staticPaths = await getStaticPaths()

    // get an array of promises
    const revalidatePaths = staticPaths.map((path) =>
          res.unstable_revalidate(`/articles/${path.params.slug}`)
    );

    // run revalidation in parallel
    await Promise.all([res.unstable_revalidate('/articles'), ...revalidatePaths]);
    
    return res.status(200).send()
  } catch (err) {
    console.error(err)
    return res.status(500).send('Revalidation Error')
  }

}
```

If you need to revalidate multiple dynamic pages, you can always create a generic `getAllStaticPaths` functions that basically just calls `getStaticPaths` from all of your pages and creates one big path array, like below:

```
import { getStaticPaths as getArticlePaths } from 'pages/articles/[slug]';
import { getStaticPaths as getContactPaths } from 'pages/contacts/[[...slug]]';
import { getStaticPaths as getCountryPaths } from 'pages/countries/[slug]';

export async function getAllStaticPaths() {
  const res = await Promise.all([getArticlePaths(), getContactPaths(), getCountryPaths()]);
  const articlePaths = res[0].paths.map((path) => `/articles/${path.params.slug}`);
  const contactPaths = res[1].paths.map((path) => `/contacts/${path.params.slug}`);
  const countryPaths = res[2].paths.map((path) => `/countries/${path.params.slug}`);

  return articlePaths.concat(contactPaths).concat(countryPaths);
}
```

Of course, this is very manual, but you can probably get fancy and automate this for all pages by dynamically importing each page as a module (using a library like `import-glob`) and extracting the `getStaticPaths` method. The hard part is prepending the proper path prefix (ie. `/articles` or `/file/[folderSlug]/[fileSlug]`). Here's some partially completed code to give you an idea:

**WARNING: This code doesn't work, it's just a demo of an idea**
```
import allPages from './pages/**/*.js` // need the import-glob library installed for this

export async function getAllStaticPathsFancyWay() {

 // grabs pages with getStaticPaths method, filtering out those that don't
  const getStaticPathMethods = allPages.map((page) => page.getStaticPaths).filter(Boolean);

 // actually get the static paths now
  const staticPaths = await Promise.all(getStaticPathMethods.map((method) => method()));

 // create one big array of paths (note that you would need to somehow get the path prefix. I was too lazy to try and figure it out.
  const flattenedPaths = staticPaths.flat().map(path => `/SOMEHOW_PREPEND_PREFIX_HERE/${path.params.slug}`);
}
```

You can get even FANCIER and rewrite the above function so you only grab paths from a subset of pages, but I'm getting carried away.

**_Or, you know, the next team can make life easy and give an option to invalidate nested pages for us :)_**

Not sure if this is exactly what you wanted but maybe it can help someone looking to revalidate dynamic pages.

**Reply by martinnabhan** [reply]
This does indeed revalidate dynamic pages's static paths,
but it doesn't revalidate dynamic paths (`fallback: 'blocking, paths; []`),
which is what I think we're all asking for.

Currently there is no way of knowing which dynamic paths have been cached (as far as I know),
which makes revalidating them impossible.

Since we can't know which dynamic paths have been cached we would need to be able to pass either `/article/[articleId]` or `/article/*` to `unstable_revalidate`, which currently isn't supported.

**Reply by kag359six** [reply]
@martinnabhan ah I see what you're saying. Maybe some modifications to the code below would work? (even though it's still a workaround and I agree it should be supported by next.js out the box):

```
    import glob from 'fast-glob'
    
    // assuming this is called from api/revalidate.js
    // returns file path using .next/server/pages as the working directory (may vary depending on build output settings)
    const files = await glob(`../articles/*.html`);
    
     // files will be something like "articles/a.js", "articles/b.js"
     // this map call turns "articles/a.js" into "/articles/a"
    const paths = files.map((file) => `/${file.replace('.html', '')}`);

    await Promise.all(paths.map(path => res.unstable_revalidate(path)));
```

The idea being that we can see which dynamic pages have been cached by poking through the `.next/server/pages` directory and searching for any generated html files.

I didn't test this though, so I could be off the mark.

**Reply by martinnabhan** [reply]
Thank you for this! I didn't know there were actual html files being saved for dynamic routes.
I've tried it out and it seems to work fine.

Unfortunately it seems like Next.js revalidation works by fetching the page with a special header, causing it to be re-rendered.
So for every page being revalidated a request is made, meaning that if you want to revalidate thousands of pages you will be sending thousands of requests to your server. I would much prefer it if the cache for the page is simply cleared, and the page gets re-rendered when a request comes in. With the current implementation I guess I'll have to keep using `getServerSideProps`, but thank you for the help!

**Comment by meotimdihia** [comment]
@[leerob](https://github.com/leerob) 

- https://symfony.com/doc/current/components/cache/cache_invalidation.html
- https://github.com/vercel/next.js/discussions/34543

We still need a lot of things to clear cache while using Cloudflare.

**Comment by j4nos** [comment]


**Reply by meotimdihia** [reply]
I don't think this question should be asked here.
If one of the array generate an error page, it will happen error, I think.
